File: paladins/src/main/java/hirez/paladins/object/Player.java
package hirez.paladins.object;

import com.fasterxml.jackson.annotation.JsonProperty;
import hirez.api.object.MergedAccount;
import hirez.api.object.adapters.DateTimeFormat;
import hirez.api.object.adapters.DurationTime;
import hirez.api.object.interfaces.ReturnedMessage;
import lombok.Data;

import java.time.Duration;
import java.time.temporal.ChronoUnit;
import java.util.Date;

@Data
public class Player implements ReturnedMessage {
    @JsonProperty("ActivePlayerId")
    private final long activePlayerId;
    @JsonProperty("AvatarId")
    private final long avatarId;
    @JsonProperty("AvatarURL")
    private final String avatarURL;
    @DateTimeFormat
    @JsonProperty("Created_Datetime")
    private final Date createdDatetime;
    @DurationTime(ChronoUnit.HOURS)
    @JsonProperty("HoursPlayed")
    private final Duration hoursPlayed;
    @JsonProperty("Id")
    private final long id;
    @DateTimeFormat
    @JsonProperty("Last_Login_Datetime")
    private final Date lastLoginDatetime;
    @JsonProperty("Leaves")
    private final int leaves;
    @JsonProperty("Level")
    private final int level;
    @JsonProperty("LoadingFrame")
    private final String loadingFrame;
    @JsonProperty("Losses")
    private final int losses;
    @JsonProperty("MasteryLevel")
    private final int masteryLevel;
    @JsonProperty("MergedPlayers")
    private final MergedAccount[] mergedPlayers;
    @DurationTime(ChronoUnit.MINUTES)
    @JsonProperty("MinutesPlayed")
    private final Duration minutesPlayed;
    @JsonProperty("Name")
    private final String name;
    @JsonProperty("Personal_Status_Message")
    private final String personalStatusMessage;
    @JsonProperty("Platform")
    private final String platform;
    @JsonProperty("RankedConquest")
    private final Rank rankedConquest;
    @JsonProperty("RankedController")
    private final Rank rankedController;
    @JsonProperty("RankedKBM")
    private final Rank rankedKBM;
    @JsonProperty("Region")
    private final String region;
    @JsonProperty("TeamId")
    private final long teamId;
    @JsonProperty("Team_Name")
    private final String teamName;
    @JsonProperty("Tier_Conquest")
    private final int tierConquest;
    @JsonProperty("Tier_RankedController")
    private final int tierRankedController;
    @JsonProperty("Tier_RankedKBM")
    private final int tierRankedKBM;
    @JsonProperty("Title")
    private final String title;
    @JsonProperty("Total_Achievements")
    private final int totalAchievements;
    @JsonProperty("Total_Worshippers")
    private final int totalWorshippers;
    @JsonProperty("Total_XP")
    private final int totalExp;
    @JsonProperty("Wins")
    private final int wins;
    @JsonProperty("hz_gamer_tag")
    private final String hiRezGamerTag;
    @JsonProperty("hz_player_name")
    private final String hiRezPlayerName;
    @JsonProperty("ret_msg")
    private final String returnedMessage;
}

File: paladins/src/main/kotlin/hirez/paladins/json/player.kt
package hirez.paladins.json

import com.google.gson.annotations.SerializedName
import hirez.json.IdObject
import hirez.json.MergedPlayer
import hirez.json.RankedItem
import hirez.json.ReturnMessage
import java.util.*

/**
 *
 * @author Damian Staszewski [damian@stachuofficial.tv]
 * @version %I%, %G%
 * @since 1.0
 */
data class Player(
			@SerializedName("ret_msg")
			override val returnMessage: String?,
			override val id: Long,
			val leaves: Long,
			val losses: Long,
			val name: String,
			val wins: Long,
			val activePlayerId: Long,
			val mergedPlayers: List<MergedPlayer>?,
			@SerializedName("Avatar_URL")
			val avatar: String,
			val hoursPlayed: Int,
			@SerializedName("Created_Datetime")
			val createdAt: Date,
			@SerializedName("Last_Login_Datetime")
			val lastLogin: Date,
			val level: Int,
			val masteryLevel: Int,
			@SerializedName("Personal_Status_Message")
			val statusMessage: String?,
			val region: String,
			val teamId: Int,
			val teamName: String,
			val totalAchievements: Int,
			val totalWorshippers: Int,
			@SerializedName("hz_gamer_tag")
			val hiRezGamerTag: String?,
			@SerializedName("hz_player_name")
			val hiRezPlayerName: String,
			val rankedConquest: RankedItem,
			val rankedController: RankedItem,
			val rankedKBM: RankedItem
) : ReturnMessage, IdObject<Long> {
	val tierConquest = rankedConquest.tier
	val tierRankedController = rankedController.tier
	val tierRankedKBM = rankedKBM.tier
}

data class ChampionLeaderboard(
			@SerializedName("champion_id")
			val championId: Long,
			val losses: Int,
			@SerializedName("player_id")
			val playerId: Long,
			@SerializedName("player_name")
			val playerName: String,
			@SerializedName("player_ranking")
			val playerRanking: Float,
			val rank: Int,
			@SerializedName("ret_msg")
			override val returnMessage: String?,
			val wins: Int
) : ReturnMessage

data class Loadout(
			@SerializedName("ChampionId")
			val championId: Long,
			@SerializedName("ChampionName")
			val championName: String,
			@SerializedName("DeckId")
			val deckId: Long,
			@SerializedName("DeckName")
			val deckName: String,
			@SerializedName("LoadoutItems")
			val items: List<CardItem>,
			val playerId: Long,
			val playerName: String,
			@SerializedName("ret_msg")
			override val returnMessage: String?
) : ReturnMessage {
	data class CardItem(
				@SerializedName("ItemId")
				override val id: Long,
				@SerializedName("ItemName")
				val name: String,
				@SerializedName("Points")
				val points: Int
	) : IdObject<Long>
}

data class PlatformPlayer(
			@SerializedName("Name")
			val name: String,
			@SerializedName("gamer_tag")
			val gamerTag: String,
			val platform: String,
			@SerializedName("player_id")
			val playerId: Long,
			@SerializedName("portal_userid")
			val portalUserid: Long,
			@SerializedName("ret_msg")
			override val returnMessage: String?
) : ReturnMessage